### PR TITLE
add: resonator

### LINKED ISSUE #426: Create <Resonator />
The resonator in our schematic symbol needs to be created with 3 pins. Take reference from how the `<Pinheader />` is implemented with multiple pins.

Image

### PR DESCRIPTION
**Pull request by Anshgrover23**
Fixes #426 
/claim #426

Note: please first merge this one: [implemented source_simple_resonator in tscircuit/circuit-json](https://github.com/tscircuit/circuit-json/pull/102)

- So that i can run tests and verify this one.

### CODE REVIEW

**Comment by Anshgrover23** [comment]
@seveibar @imrishabh18 the checks are currently failing because we have not merged that one PR, kindly merge that one first.

File: lib/components/normal-components/Resonator.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import { resonatorProps } from "@tscircuit/props"
+import { NormalComponent } from "../base-components/NormalComponent"
+import { Port } from "../primitive-components/Port"
+import { SourceSimpleResonator } from "circuit-json"
+import type { SchematicPortArrangement } from "@tscircuit/props"
+
+export class Resonator extends NormalComponent<typeof resonatorProps> {
+  get config() {
+    return {
```
Comment (imrishabh18) [inline lib/components/normal-components/Resonator.ts:9]: It is missing the default `schematicSymbolName` property.

File: lib/components/normal-components/Resonator.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      leftSize: 1,
+      rightSize: 2,
+    }
+  }
+
+  doInitialSourceRender() {
+    const { db } = this.root!
+    const { _parsedProps: props } = this
+    const source_component = db.source_component.insert({
+      ftype: "resonator",
```
Comment (imrishabh18) [inline lib/components/normal-components/Resonator.ts:50]: The ftype value should be the one used in the circuit json, so it should be `simple_resonator`

**Review by imrishabh18 — CHANGES_REQUESTED** [review]

**Comment by imrishabh18** [comment]
@Anshgrover23 You don't have to wait for us to merge the circuit-json PR. You can test it in your local by using the `yalc`package

File: lib/components/normal-components/Resonator.ts (hunk 1 above)
Comment (Anshgrover23) [reply lib/components/normal-components/Resonator.ts:10]: done.

File: lib/components/normal-components/Resonator.ts (hunk 2 above)
Comment (Anshgrover23) [reply lib/components/normal-components/Resonator.ts]: done.

File: lib/components/normal-components/Resonator.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  _getImpliedFootprintString(): string | null {
+    const frequency = this._parsedProps.frequency
+    const loadCapacitance = this._parsedProps.loadCapacitance
+    if (frequency && loadCapacitance) {
+      return `resonator_3pin_f${frequency}_cl${loadCapacitance}`
+    }
+    return null
+  }
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:26]:
```suggestion
```

File: lib/components/normal-components/Resonator.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    for (let i = 1; i <= pinCount; i++) {
+      this.add(
+        new Port({
+          name: `pin${i}`,
+          pinNumber: i,
+          aliases: [],
+        }),
+      )
+    }
+  }
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:39]: can you try removing this? does it work without it?

Comment (seveibar) [reply lib/components/normal-components/Resonator.ts:39]: i could be wrong

File: lib/components/normal-components/Resonator.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  _getSchematicPortArrangement(): SchematicPortArrangement | null {
+    const pinCount = 3 // Fixed for 3-pin resonator
+    return {
+      leftSize: 1,
+      rightSize: 2,
+    }
+  }
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:47]:
```suggestion
```

File: lib/components/normal-components/Resonator.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  doInitialSourceRender() {
+    const { db } = this.root!
+    const { _parsedProps: props } = this
+    const source_component = db.source_component.insert({
+      ftype: "simple_resonator",
+      name: props.name,
+      frequency: props.frequency,
+      load_capacitance: props.loadCapacitance,
+      pin_count: 3,
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:57]:
```suggestion
```

File: lib/components/normal-components/Resonator.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+export class Resonator extends NormalComponent<typeof resonatorProps> {
+  get config() {
+    return {
+      componentName: "Resonator",
+      schematicSymbolName:
+        this.props.symbolName ?? ("resonator" as BaseSymbolName),
+      zodProps: resonatorProps, // Using resonatorProps for validation
+      shouldRenderAsSchematicBox: true,
+      pinVariant: this.props.pinVariant,
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:14]: why is this here?

File: lib/components/normal-components/Resonator.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import { NormalComponent } from "../base-components/NormalComponent"
+import { type SourceSimpleResonator } from "circuit-json"
+import type { BaseSymbolName } from "lib/utils/constants"
+
+export class Resonator extends NormalComponent<typeof resonatorProps> {
+  get config() {
+    return {
+      componentName: "Resonator",
+      schematicSymbolName:
+        this.props.symbolName ?? ("resonator" as BaseSymbolName),
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:11]: choose the symbol based on the pin varient

**Review by seveibar — CHANGES_REQUESTED** [review]
see comments

File: lib/components/normal-components/Resonator.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import { resonatorProps } from "@tscircuit/props"
+import { NormalComponent } from "../base-components/NormalComponent"
+import type { BaseSymbolName } from "lib/utils/constants"
+
+export class Resonator extends NormalComponent<typeof resonatorProps> {
+  get config() {
+    const getSymbolName = (variant: string | undefined): BaseSymbolName => {
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:7]: don't nest functions unnecessarily, move the function outside of the `get config() {` function

why? it simplifies nesting code and allows you to use multiple files.

You should rename the function `getResonatorSymbolName` but you can keep it in this file since it's already so short

**Review by seveibar — COMMENTED** [review]
see my comment, we can merge after that adjustment

File: lib/components/normal-components/Resonator.ts (hunk 9 above)
Comment (seveibar) [reply lib/components/normal-components/Resonator.ts:7]: CC @Abse2001 this is a good lint rule to add, this pattern isn't allowed on any project

I think eslint has a version of it although we'll want to be careful not to break e.g. react

https://eslint.org/docs/latest/rules/no-inner-declarations

image

Comment (Anshgrover23) [reply lib/components/normal-components/Resonator.ts]: done.

File: lib/components/normal-components/Resonator.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { resonatorProps } from "@tscircuit/props"
+import { NormalComponent } from "../base-components/NormalComponent"
+import type { BaseSymbolName } from "lib/utils/constants"
+
+export class Resonator extends NormalComponent<typeof resonatorProps> {
+  getResonatorSymbolName = (variant: string | undefined): BaseSymbolName => {
```
Comment (seveibar) [inline lib/components/normal-components/Resonator.ts:6]: this is even worse! it needs to be in the root scope. You've added a new public property to resonators.

You could theoretically do this if it was internal, but then you'd have to change the function

**Review by seveibar — CHANGES_REQUESTED** [review]

File: tests/components/normal-components/resonator.test.tsx [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import { it, expect, describe } from "bun:test"
+import { getTestFixture } from "tests/fixtures/get-test-fixture"
+
+describe("Resonator Component", () => {
+  it("should render a resonator with no ground pins", async () => {
+    const { circuit } = getTestFixture()
+    circuit.add(
+      <board width="10mm" height="10mm">
+        <resonator
+          name="R1"
```
Comment (seveibar) [inline tests/components/normal-components/resonator.test.tsx:10]: only resistors should use R as the reference designator

```suggestion
          name="K1"
```

File: tests/components/normal-components/resonator.test.tsx [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    circuit.render()
+    expect(circuit).toMatchSchematicSnapshot("resonator-single-ground")
+  })
+
+  it("should render a resonator with two ground pins", async () => {
+    const { circuit } = getTestFixture()
+    circuit.add(
+      <board width="10mm" height="10mm">
+        <resonator
+          name="R3"
```
Comment (seveibar) [inline tests/components/normal-components/resonator.test.tsx:42]:
```suggestion
          name="K3"
```

File: tests/components/normal-components/resonator.test.tsx [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    )
+    circuit.render()
+    expect(circuit).toMatchSchematicSnapshot("resonator-two-ground")
+  })
+
+  it("should render a resonator without pinVariant specified", async () => {
+    const { circuit } = getTestFixture()
+    circuit.add(
+      <board width="10mm" height="10mm">
+        <resonator name="R4" frequency="8MHz" loadCapacitance="15pF" />
```
Comment (seveibar) [inline tests/components/normal-components/resonator.test.tsx:57]:
```suggestion
        <resonator name="K4" frequency="8MHz" loadCapacitance="15pF" />
```

**Review by seveibar — APPROVED** [review]
ping me when you have the new reference designators and you're good to go 👍

**Comment by seveibar** [comment]
@Anshgrover23 make sure to fix merge conflicts

**Review by seveibar — APPROVED** [review]
nice work!
